Victor CHAU & Tahar AMAIRI
# IFT 870 - Forage de données
## TP2 : Prétraitement et représentation de données

### Sommaire

### TODO

## Librairies

In [1]:
# import libs
from matplotlib import pyplot as plt
from IPython.display import display
import seaborn as sns
import pandas as pd
import numpy as np

## Importation des données

In [2]:
# read CSVs
dfProd = pd.read_csv('./product2.csv', sep=';', encoding='latin1')
dfPack = pd.read_csv('./package2.csv', sep=';', encoding='latin1')

## Question 1

In [3]:
def desc_data(name, data):
    nbuniques = data.nunique()
    nb_m = data.isnull().sum()
    ratio_m = (data.isnull().sum()/data.shape[0])
    manquant = pd.concat([nb_m, ratio_m], axis=1, sort=False)
    frq =     frq = data.mode(dropna=False)
    df = pd.DataFrame({'Types': data[list(manquant.index.values)].dtypes,
                       'Nb_Uniques': nbuniques,
                       'Nb manquants': nb_m,
                      'Ratio manquants%': ratio_m*100,
                      'Valeur la plus fréquente': list(frq)})
    print("Table",name)
    print(df)
    print("")

### Table Product2

In [4]:
# print
display(dfProd.head())

PRODUCTID PRODUCTNDC          PRODUCTTYPENAME  PROPRIETARYNAME  \
0       NaN  0002-0800           HUMAN OTC DRUG  Sterile Diluent   
1       NaN  0002-1200  HUMAN PRESCRIPTION DRUG           Amyvid   
2       NaN  0002-1433  HUMAN PRESCRIPTION DRUG        Trulicity   
3       NaN  0002-1434  HUMAN PRESCRIPTION DRUG        Trulicity   
4       NaN  0002-1436  HUMAN PRESCRIPTION DRUG         EMGALITY   

  PROPRIETARYNAMESUFFIX NONPROPRIETARYNAME       DOSAGEFORMNAME     ROUTENAME  \
0                   NaN            diluent  INJECTION, SOLUTION  SUBCUTANEOUS   
1                   NaN   Florbetapir F 18  INJECTION, SOLUTION   INTRAVENOUS   
2                   NaN        Dulaglutide  INJECTION, SOLUTION  SUBCUTANEOUS   
3                   NaN        Dulaglutide  INJECTION, SOLUTION  SUBCUTANEOUS   
4                   NaN       galcanezumab  INJECTION, SOLUTION  SUBCUTANEOUS   

   STARTMARKETINGDATE  ENDMARKETINGDATE MARKETINGCATEGORYNAME  \
0            19870710               NaN                   NDA   
1            20120601               NaN                   NDA   
2            20140918               NaN                   BLA   
3            20140918               NaN                   BLA   
4            20180927               NaN                   BLA   

  APPLICATIONNUMBER LABELERNAME     SUBSTANCENAME ACTIVE_NUMERATOR_STRENGTH  \
0         NDA018781          10             WATER                         1   
1         NDA202008          10  FLORBETAPIR F-18                        51   
2         BLA125469          10       DULAGLUTIDE                       .75   
3         BLA125469          10       DULAGLUTIDE                       1.5   
4         BLA761063          10      GALCANEZUMAB                       120   

  ACTIVE_INGRED_UNIT                                      PHARM_CLASSES  \
0              mL/mL                                                NaN   
1             mCi/mL  Radioactive Diagnostic Agent [EPC],Positron Em...   
2            mg/.5mL  GLP-1 Receptor Agonist [EPC],Glucagon-Like Pep...   
3            mg/.5mL  GLP-1 Receptor Agonist [EPC],Glucagon-Like Pep...   
4              mg/mL                                                NaN   

  DEASCHEDULE NDC_EXCLUDE_FLAG  LISTING_RECORD_CERTIFIED_THROUGH  
0         NaN                N                        20201231.0  
1         NaN                N                        20211231.0  
2         NaN                N                        20201231.0  
3         NaN                N                        20201231.0  
4         NaN                N                        20201231.0

In [5]:
#number of features
product_num_cols = len(dfProd.columns)
print(f"Number of columns in product2.csv: {product_num_cols}")

Number of columns in product2.csv: 20


In [6]:
desc_data("Product2", dfProd)

Table Product2
                                    Types  Nb_Uniques  Nb manquants  \
PRODUCTID                          object       91678          1560   
PRODUCTNDC                         object       91468             0   
PRODUCTTYPENAME                    object           7             0   
PROPRIETARYNAME                    object       32716             6   
PROPRIETARYNAMESUFFIX              object        4022         83075   
NONPROPRIETARYNAME                 object       16257             4   
DOSAGEFORMNAME                     object         134             0   
ROUTENAME                          object         180          1932   
STARTMARKETINGDATE                  int64        7262             0   
ENDMARKETINGDATE                  float64         742         88915   
MARKETINGCATEGORYNAME              object          10             0   
APPLICATIONNUMBER                  object       10711         13097   
LABELERNAME                        object        6611         

In [7]:
for col in dfProd.columns:
    print(f'Column: {col}')
    print(f'Value counts: \n{dfProd[col].value_counts()}')
    print('\n')

Column: PRODUCTID
Value counts: 
0002-3251_67a53369-eead-4f2c-afe9-f3274899c47e    1
63868-061_95209f78-78d1-3272-e053-2995a90a311a    1
63868-086_4e8dbf4d-856e-45d7-b993-bcc4939fc2de    1
63868-084_94592969-5251-a13e-e053-2995a90ab182    1
63868-083_7e490b7b-e78e-5917-e053-2991aa0a3d17    1
                                                 ..
46287-014_451b0304-847a-41b2-a5d6-8574ce414148    1
46287-012_9d3dbf04-dbc9-47d3-855f-96ca1cf0788b    1
46287-009_75907d04-0c9d-4a6a-b8c0-7d88ab5ad337    1
46287-006_ab06fa37-b637-4323-b95f-de1a8958c5b6    1
99207-850_76ac026f-e6f1-4d1f-8144-9b5492e2d1bb    1
Name: PRODUCTID, Length: 91678, dtype: int64


Column: PRODUCTNDC
Value counts: 
OTC MONOGRAPH FINAL        111
UNAPPROVED HOMEOPATHIC     109
ANDA                       105
OTC MONOGRAPH NOT FINAL    101
NDA                         60
                          ... 
44911-0125                   1
44911-0120                   1
44911-0119                   1
44911-0117                   1
9920

`NONPROPRIETARYNAME`, `DOSAGEFORMNAME`, `ROUTENAME`, `SUBSTANCENAME`, `ACTIVE_NUMERATOR_STRENGTH`, `ACTIVE_INGRED_UNIT`, `PHARM_CLASSES` contiennent plusieurs mots/nombres distincts reliés par une virgule ou un point-virgule.

### Table Package2

In [8]:
display(dfPack.head())

PRODUCTID PRODUCTNDC NDCPACKAGECODE  \
0  0002-0800_94c48759-29bb-402d-afff-9a713be11f0e  0002-0800   0002-0800-01   
1  0002-1200_35551a38-7a8d-43b8-8abd-f6cb7549e932  0002-1200   0002-1200-30   
2  0002-1200_35551a38-7a8d-43b8-8abd-f6cb7549e932  0002-1200   0002-1200-50   
3  0002-1433_42a80046-fd68-4b80-819c-a443b7816edb  0002-1433   0002-1433-61   
4  0002-1433_42a80046-fd68-4b80-819c-a443b7816edb  0002-1433   0002-1433-80   

                                  PACKAGEDESCRIPTION  STARTMARKETINGDATE  \
0  1 VIAL in 1 CARTON (0002-0800-01)  > 10 mL in ...            19870710   
1  1 VIAL, MULTI-DOSE in 1 CAN (0002-1200-30)  > ...            20120601   
2  1 VIAL, MULTI-DOSE in 1 CAN (0002-1200-50)  > ...            20120601   
3  2 SYRINGE in 1 CARTON (0002-1433-61)  > .5 mL ...            20141107   
4  4 SYRINGE in 1 CARTON (0002-1433-80)  > .5 mL ...            20141107   

   ENDMARKETINGDATE NDC_EXCLUDE_FLAG SAMPLE_PACKAGE  
0               NaN                N              N  
1               NaN                N              N  
2               NaN                N              N  
3               NaN                N              Y  
4               NaN                N              N

In [9]:
#number of features
package_num_cols = len(dfPack.columns)
print(f"Number of columns in package2.csv: {package_num_cols}")

Number of columns in package2.csv: 8


In [10]:
desc_data("Package2", dfPack)

Table Package2
                      Types  Nb_Uniques  Nb manquants  Ratio manquants%  \
PRODUCTID            object       93084             0          0.000000   
PRODUCTNDC           object       91080          1500          0.862629   
NDCPACKAGECODE       object      171447          2346          1.349152   
PACKAGEDESCRIPTION   object      173885             0          0.000000   
STARTMARKETINGDATE    int64        7401             0          0.000000   
ENDMARKETINGDATE    float64         767        167431         96.287244   
NDC_EXCLUDE_FLAG     object           1             0          0.000000   
SAMPLE_PACKAGE       object           2             0          0.000000   

                   Valeur la plus fréquente  
PRODUCTID                         PRODUCTID  
PRODUCTNDC                       PRODUCTNDC  
NDCPACKAGECODE               NDCPACKAGECODE  
PACKAGEDESCRIPTION       PACKAGEDESCRIPTION  
STARTMARKETINGDATE       STARTMARKETINGDATE  
ENDMARKETINGDATE           ENDMAR

In [11]:
for col in dfPack.columns:
    print(f'Column: {col}')
    print(f'Value counts: \n{dfPack[col].value_counts()}')
    print('\n')

Column: PRODUCTID
Value counts: 
73069-525_e338a0c8-2ca2-42a4-8e4d-d76b4002ee48     72
73069-037_9eff0917-36c9-4fee-8799-233396692ca3     72
73069-277_c983c583-ed08-4a46-9cee-92a384658216     72
73069-500_88c8e558-be0a-4661-8742-1d15e919e113     72
73069-450_44474104-bd9a-403b-b2fd-0a47cac17707     72
                                                   ..
58232-0040_9ea0d32f-a352-4c4d-b294-e619e2c2f9e2     1
58232-0054_12919a4d-c0c3-4a43-9cbc-28650027235f     1
22840-1469_94674694-b4db-3a94-e053-2a95a90a36d5     1
58232-0063_503ca461-aa42-4dc2-bc8f-eada3ada01f6     1
0002-0800_94c48759-29bb-402d-afff-9a713be11f0e      1
Name: PRODUCTID, Length: 93084, dtype: int64


Column: PRODUCTNDC
Value counts: 
73069-206     72
73069-300     72
73069-035     72
73069-024     72
73069-249     72
              ..
50090-4082     1
50090-4083     1
50090-4100     1
50090-4102     1
55301-038      1
Name: PRODUCTNDC, Length: 91080, dtype: int64


Column: NDCPACKAGECODE
Value counts: 
20211201        43


## Question 2

1. `PRODUCTID`, `PRODUCTNDC` et `NDCPACKAGECODE` ont une valeur qui commence par la même base : un segment du code de l'étiqueteur et du code du produit du numéro du code national des médicaments, séparés par un trait d'union.
<br/>`PRODUCTID` et `NDCPACKAGECODE` doivent au moins contenir la valeur de `PRODUCTNDC`.

2. `STARTMARKETINGDATE` est la date du début de la commercialisation du médicament; et`ENDMARKETINGDATE` est la date où le produit ne sera plus disponible sur le marché. La date du début `STARTMARKETINGDATE` doit être avant la date de fin `ENDMARKETINGDATE`.

## Question 3

### vérification règle 1 de la question2

**table product2**

In [12]:
#vérification règle 1 de la question2
compt=0
for index, row in dfProd.iterrows():
    if (str(row['PRODUCTNDC']) not in str(row['PRODUCTID'])) and pd.notna(row['PRODUCTID']) and pd.notna(row['PRODUCTNDC']):
        print(row['PRODUCTNDC'],row['PRODUCTID'])
        dfProd.loc[index, 'PRODUCTNDC'] = np.nan #forcer ces valeurs à NaN
        compt+=1

05-juin 0006-0005_0c7a3452-ecb2-4f66-ad52-94f8eaf8cde8
19-juin 0006-0019_54e9c31a-9429-4842-b2d6-0cc1e5ad613c
03-sept 0009-0003_67759a7c-ea06-4151-87e1-a301c44d67cd
05-sept 0009-0005_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
11-sept 0009-0011_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
12-sept 0009-0012_8e4b06b5-7335-49ae-bb11-1a4195716664
13-sept 0009-0013_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
16-sept 0009-0016_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
17-sept 0009-0017_12eb5556-e1c1-468d-a841-2b240ccc8ddf
18-sept 0009-0018_67759a7c-ea06-4151-87e1-a301c44d67cd
20-sept 0009-0020_8f9c2163-1d7f-4e6e-bbe5-d758de474744
22-sept 0009-0022_8f9c2163-1d7f-4e6e-bbe5-d758de474744
29-sept 0009-0029_1ccf0901-e87d-4b6c-b4cc-7d6bada4724d
OTC MONOGRAPH NOT FINAL 53603-1049_464b7a83-0be4-6d33-e054-00144ff88e88
NDA 53622-001_5bc4566c-cf41-4690-ba5b-32e49cec5c7d
OTC MONOGRAPH NOT FINAL 53633-100_c2c9702b-1ecf-49bc-b9a5-bdc02ce9c05d
OTC MONOGRAPH FINAL 53634-010_2826812a-5562-418f-a5f3-67c113c7c669
OTC MONOGRAPH FINAL 5363

In [13]:
print('Nombre de valeurs incohérentes: ',compt)

Nombre de valeurs incohérentes:  513


>On remarque dans la table _Product2_ que des mauvaises valeurs ont été mises dans l'attribut `PRODUCTNDC`

**table package2**

In [14]:
#vérification règle 1 de la question2
compt=0
for index, row in dfPack.iterrows():
    if (str(row['PRODUCTNDC']) not in str(row['PRODUCTID'])) and pd.notna(row['PRODUCTID']) and pd.notna(row['PRODUCTNDC']):
        print(row['PRODUCTNDC'],row['PRODUCTID'])
        dfPack.loc[index, 'PRODUCTNDC'] = np.nan#forcer ces valeurs à NaN
        compt+=1

05-juin 0006-0005_0c7a3452-ecb2-4f66-ad52-94f8eaf8cde8
05-juin 0006-0005_0c7a3452-ecb2-4f66-ad52-94f8eaf8cde8
19-juin 0006-0019_54e9c31a-9429-4842-b2d6-0cc1e5ad613c
03-sept 0009-0003_67759a7c-ea06-4151-87e1-a301c44d67cd
05-sept 0009-0005_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
11-sept 0009-0011_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
12-sept 0009-0012_8e4b06b5-7335-49ae-bb11-1a4195716664
13-sept 0009-0013_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
13-sept 0009-0013_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
16-sept 0009-0016_c9aa26c1-05c3-479c-90eb-63b2181c5e7e
17-sept 0009-0017_12eb5556-e1c1-468d-a841-2b240ccc8ddf
17-sept 0009-0017_12eb5556-e1c1-468d-a841-2b240ccc8ddf
18-sept 0009-0018_67759a7c-ea06-4151-87e1-a301c44d67cd
20-sept 0009-0020_8f9c2163-1d7f-4e6e-bbe5-d758de474744
22-sept 0009-0022_8f9c2163-1d7f-4e6e-bbe5-d758de474744
29-sept 0009-0029_1ccf0901-e87d-4b6c-b4cc-7d6bada4724d
29-sept 0009-0029_1ccf0901-e87d-4b6c-b4cc-7d6bada4724d
29-sept 0009-0029_1ccf0901-e87d-4b6c-b4cc-7d6bada4724d
29-sept 00

In [15]:
print('Nombre de valeurs incohérentes dans PRODUCTNDC: ',compt)

Nombre de valeurs incohérentes dans PRODUCTNDC:  519


>On remarque dans la table _Package2_ que des dates ont été mises dans l'attribut `PRODUCTNDC`

In [16]:
#vérification règle 1 de la question2
compt=0
for index, row in dfPack.iterrows():
    if (str(row['PRODUCTNDC']) not in str(row['NDCPACKAGECODE'])) and pd.notna(row['NDCPACKAGECODE']) and pd.notna(row['PRODUCTNDC']):
        print(row['PRODUCTNDC'],row['NDCPACKAGECODE'],row['PRODUCTID'])
        dfPack.loc[index, 'NDCPACKAGECODE']= np.nan #forcer ces valeurs à NaN
        compt+=1

0220-2997 0220-1997-41 0220-2997_8590fcc9-c0c3-f43b-e053-2991aa0a692e
37000-005 20211201 37000-005_99fe4eef-411e-08db-e053-2995a90a295e
37000-005 20211201 37000-005_99fe4eef-411e-08db-e053-2995a90a295e
37000-005 20211201 37000-005_99fe4eef-411e-08db-e053-2995a90a295e
37000-005 20211201 37000-005_99fe4eef-411e-08db-e053-2995a90a295e
37000-005 20211201 37000-005_99fe4eef-411e-08db-e053-2995a90a295e
37000-005 20211201 37000-005_99fe4eef-411e-08db-e053-2995a90a295e
37000-008 20211201 37000-008_99ed1c42-5ea3-e2a1-e053-2a95a90a304a
37000-008 20211201 37000-008_99ed1c42-5ea3-e2a1-e053-2a95a90a304a
37000-008 20211201 37000-008_99ed1c42-5ea3-e2a1-e053-2a95a90a304a
37000-022 20211201 37000-022_9998d1b8-fdd6-21d2-e053-2a95a90a3a92
37000-022 20211201 37000-022_9998d1b8-fdd6-21d2-e053-2a95a90a3a92
37000-022 20211201 37000-022_9998d1b8-fdd6-21d2-e053-2a95a90a3a92
37000-024 20211201 37000-024_9a114686-42da-4f8e-e053-2995a90ae09c
37000-025 20211201 37000-025_99eb52b6-2a12-b2dc-e053-2a95a90a0f53
37000-

In [17]:
print('Nombre de valeurs incohérentes dans NDCPACKAGECODE: ',compt)

Nombre de valeurs incohérentes dans NDCPACKAGECODE:  155


>On remarque dans la table _Product2_ que des dates ont été mises dans l'attribut `NDCPACKAGECODE`
<br/> On voit aussi une erreur de frappe : 0220-`2`997 et 0220-`1`997-41 et en vérifiant le code `PRODUCTID`, on se rend compte que `NDCPACKAGECODE` aurait du être 0220-`2`997-41

### vérification règle 2 de la question2

**table package2**

In [18]:
#vérification règle 2 de la question2

#Dates manquantes avant changement de format
print("Dates de début manquantes avant changement de format: ",dfPack['STARTMARKETINGDATE'].isna().sum())
print("Dates de fin manquantes avant changement de format: ",dfPack['ENDMARKETINGDATE'].isna().sum())
#Changement de format plus 
dfPack['STARTMARKETINGDATE']= pd.to_datetime(dfPack['STARTMARKETINGDATE'], format='%Y%m%d', errors='coerce')
dfPack['ENDMARKETINGDATE']= pd.to_datetime(dfPack['ENDMARKETINGDATE'], format='%Y%m%d', errors='coerce')
#Dates manquantes après changement de format
print("Dates de début manquantes après changement de format: ",dfPack['STARTMARKETINGDATE'].isna().sum())
print("Dates de fin manquantes après changement de format: ",dfPack['ENDMARKETINGDATE'].isna().sum())

Dates de début manquantes avant changement de format:  0
Dates de fin manquantes avant changement de format:  167431
Dates de début manquantes après changement de format:  243
Dates de fin manquantes après changement de format:  167431


In [19]:
dfPack['STARTMARKETINGDATE'].where(dfPack['STARTMARKETINGDATE']>=dfPack['ENDMARKETINGDATE']).dropna()

Series([], Name: STARTMARKETINGDATE, dtype: datetime64[ns])

>Il y a 243 dates qui n'étaient dans un format valide et il n'y a pas de dates qui ne sont pas cohérentes.

**table product2**

In [20]:
#vérification règle 2 de la question2

#Dates manquantes avant changement de format
print("Dates de début manquantes avant changement de format: ",dfProd['STARTMARKETINGDATE'].isna().sum())
print("Dates de fin manquantes avant changement de format: ",dfProd['ENDMARKETINGDATE'].isna().sum())
#Changement de format plus 
dfProd['STARTMARKETINGDATE']= pd.to_datetime(dfProd['STARTMARKETINGDATE'], format='%Y%m%d', errors='coerce')
dfProd['ENDMARKETINGDATE']= pd.to_datetime(dfProd['ENDMARKETINGDATE'], format='%Y%m%d', errors='coerce')
#Dates manquantes après changement de format
print("Dates de début manquantes après changement de format: ",dfProd['STARTMARKETINGDATE'].isna().sum())
print("Dates de fin manquantes après changement de format: ",dfProd['ENDMARKETINGDATE'].isna().sum())

Dates de début manquantes avant changement de format:  0
Dates de fin manquantes avant changement de format:  88915
Dates de début manquantes après changement de format:  0
Dates de fin manquantes après changement de format:  88916


In [21]:
dfProd['STARTMARKETINGDATE'].where(dfProd['STARTMARKETINGDATE']>=dfProd['ENDMARKETINGDATE']).dropna()

Series([], Name: STARTMARKETINGDATE, dtype: datetime64[ns])

>Il y a 1 date de fin qui n'était dans un format valide et il n'y a pas de dates qui ne sont pas cohérentes.

## Question 4

DOSAGEFORMNAME, ROUTENAME, SUBSTANCENAME, ACTIVE_NUMERATOR_STRENGTH, ACTIVE_INGRED_UNIT, PHARM_CLASSES contiennent plusieurs mots/nombres distincts reliés par une virgule ou un point-virgule)

In [22]:
dfProd.isnull().sum()

PRODUCTID                            1560
PRODUCTNDC                            513
PRODUCTTYPENAME                         0
PROPRIETARYNAME                         6
PROPRIETARYNAMESUFFIX               83075
NONPROPRIETARYNAME                      4
DOSAGEFORMNAME                          0
ROUTENAME                            1932
STARTMARKETINGDATE                      0
ENDMARKETINGDATE                    88916
MARKETINGCATEGORYNAME                   0
APPLICATIONNUMBER                   13097
LABELERNAME                             0
SUBSTANCENAME                        2309
ACTIVE_NUMERATOR_STRENGTH            2309
ACTIVE_INGRED_UNIT                   2309
PHARM_CLASSES                       50984
DEASCHEDULE                         88815
NDC_EXCLUDE_FLAG                        0
LISTING_RECORD_CERTIFIED_THROUGH     4325
dtype: int64

In [23]:
dfPack.isnull().sum()

PRODUCTID                  0
PRODUCTNDC              2019
NDCPACKAGECODE          2501
PACKAGEDESCRIPTION         0
STARTMARKETINGDATE       243
ENDMARKETINGDATE      167431
NDC_EXCLUDE_FLAG           0
SAMPLE_PACKAGE             0
dtype: int64

>**Idées de méthodes pour valeurs manquantes**

`PRODUCTID` : Nous pouvons soit trouver les valeurs manquantes comparant le segment de code du PRODUCTNDC correspondant, mais il nous manquerait une partie de code (SPL documentID d'après le site). Le ratio de valeurs manquantes est peu significatif donc nous décidons de drop les lignes.

In [24]:
dfProd.dropna(subset=['PRODUCTID'],inplace=True)

`PRODUCTNDC` : nous pouvons essayer de compléter les valeurs PRODUCTNDC manquantes en recherchant des valeurs PRODUCTID ou NDCPACKAGECODE correspondantes.

In [25]:
#dfProd
#for index, row in dfProd.iterrows():
 #   if dfProd[index,'PRODUCTNDC'].isna():
  #      dfProd[index,'PRODUCTNDC'].fillna(row['PRODUCTID'].astype(str).apply(lambda x: '-'.join(x.split('-')[:2])if pd.notna(x) else pd.NaN))

In [26]:
#dfPack


`PROPRIETARYNAME` : C'est le nom choisi par le propriétaire. Le nombre de valeurs manquante est minime, on peut drop les lignes.

In [27]:
dfProd.dropna(subset=['PROPRIETARYNAME'],inplace=True)

`PROPRIETARYNAMESUFFIX` : Comme cette colonne contient des suffixes pour les noms de propriétaires, les valeurs manquantes peuvent être remplies par des chaînes vides.

In [28]:
dfProd['PROPRIETARYNAMESUFFIX'] = dfProd['PROPRIETARYNAMESUFFIX'].replace(np.nan, '')

`NONPROPRIETARYNAME` : Le nombre de valeurs manquante est minime, on peut drop les lignes.

In [29]:
dfProd.dropna(subset=['NONPROPRIETARYNAME'],inplace=True)

`STARTMARKETINGDATE` : Nous allons supprimer les valeurs car le nombre est manquant est minime.

In [30]:
dfPack.dropna(subset=['STARTMARKETINGDATE'],inplace=True)

`ENDMARKETINGDATE` : si les valeurs manquantes de ces colonnes ne peuvent être trouvées en recherchant le produit avec un PRODUCTNDC correspondant ou d'autres informations d'identification, nous pouvons supposer que le produit est toujours commercialisé s'il n'a pas été explicitement abandonné.

`APPLICATIONNUMBER` : Pour les médicaments non approuvés, les valeurs restent NULL.

`SUBSTANCENAME` : Liste des produits actifs. Nous ne pouvons pas remplacer ces valeurs donc nous allons supprimer les lignes

In [31]:
dfProd.dropna(subset=['SUBSTANCENAME'],inplace=True)

`ACTIVE_NUMERATOR_STRENGTH` et `ACTIVE_INGRED_UNIT` : nous pouvo@ns supposer que les valeurs manquantes pour ces colonnes sont les mêmes, et les mêmes que SUBSTANCENAME. Nous allons les supprimer.

In [32]:
dfProd.dropna(subset=['ACTIVE_NUMERATOR_STRENGTH','ACTIVE_INGRED_UNIT'],inplace=True)

`PHARM_CLASSES` : Beaucoup de valeurs sont manquantes ici. Cela correpond aux classes pharcologiques correpondant aux SUBSTANCENAME au dessus. De plus l'attribut comportent plusieurs valeurs donc c'est compliqué.

`DEASCHEDULE` : Les valeurs CI, CII, CIII, CIV, and CV. Mais ils manquent tellement de valeurs que nous ne savons pas s'il faut les remplacer.

`LISTING_RECORD_CERTIFIED_THROUGH` : Il s'agit de la date à laquelle l'enregistrement de l'inscription expirera s'il n'est pas mis à jour ou certifié par l'entreprise. Il y en a 2. Nous prendrons la valeur la plus fréquente.

In [33]:
dfProd['LISTING_RECORD_CERTIFIED_THROUGH'] = dfProd['LISTING_RECORD_CERTIFIED_THROUGH'].replace(np.nan, dfProd['LISTING_RECORD_CERTIFIED_THROUGH'].value_counts().idxmax())

`NDCPACKAGECODE` :

In [34]:
#dfPack


## Question 5

In [35]:
dfProd.drop_duplicates(subset=['PRODUCTID'], keep="first", inplace=False)

PRODUCTID PRODUCTNDC  \
20     0002-3251_67a53369-eead-4f2c-afe9-f3274899c47e  0002-3251   
21     0002-3270_06e2a1f2-459c-45aa-9341-54e36f7726a7  0002-3270   
22     0002-4112_d561034d-ea58-45fe-9d07-2e9eba98c2e4  0002-4112   
23     0002-4115_d561034d-ea58-45fe-9d07-2e9eba98c2e4  0002-4115   
24     0002-4116_d561034d-ea58-45fe-9d07-2e9eba98c2e4  0002-4116   
...                                               ...        ...   
93233  99207-465_7578e84a-41ed-498d-8c2b-56a9931679db  99207-465   
93234  99207-466_7578e84a-41ed-498d-8c2b-56a9931679db  99207-466   
93235  99207-467_7578e84a-41ed-498d-8c2b-56a9931679db  99207-467   
93236  99207-525_d47eda34-3952-463c-9597-4225a19dbf13  99207-525   
93237  99207-850_76ac026f-e6f1-4d1f-8144-9b5492e2d1bb  99207-850   

               PRODUCTTYPENAME PROPRIETARYNAME PROPRIETARYNAMESUFFIX  \
20     HUMAN PRESCRIPTION DRUG       Strattera                         
21     HUMAN PRESCRIPTION DRUG        Cymbalta                         
22     HUMAN PRESCRIPTION DRUG         Zyprexa                         
23     HUMAN PRESCRIPTION DRUG         Zyprexa                         
24     HUMAN PRESCRIPTION DRUG         Zyprexa                         
...                        ...             ...                   ...   
93233  HUMAN PRESCRIPTION DRUG         Solodyn                         
93234  HUMAN PRESCRIPTION DRUG         Solodyn                         
93235  HUMAN PRESCRIPTION DRUG         Solodyn                         
93236  HUMAN PRESCRIPTION DRUG           Vanos                         
93237  HUMAN PRESCRIPTION DRUG            Luzu                         

              NONPROPRIETARYNAME                         DOSAGEFORMNAME  \
20     Atomoxetine hydrochloride                                CAPSULE   
21      Duloxetine hydrochloride               CAPSULE, DELAYED RELEASE   
22                    Olanzapine                                 TABLET   
23                    Olanzapine                                 TABLET   
24                    Olanzapine                                 TABLET   
...                          ...                                    ...   
93233  minocycline hydrochloride  TABLET, FILM COATED, EXTENDED RELEASE   
93234  minocycline hydrochloride  TABLET, FILM COATED, EXTENDED RELEASE   
93235  minocycline hydrochloride  TABLET, FILM COATED, EXTENDED RELEASE   
93236               fluocinonide                                  CREAM   
93237               LULICONAZOLE                                  CREAM   

      ROUTENAME STARTMARKETINGDATE ENDMARKETINGDATE MARKETINGCATEGORYNAME  \
20         ORAL         2005-02-14              NaT                   NDA   
21         ORAL         2010-01-15              NaT                   NDA   
22         ORAL         1997-06-23              NaT                   NDA   
23         ORAL         1996-10-01              NaT                   NDA   
24         ORAL         1996-10-01              NaT                   NDA   
...         ...                ...              ...                   ...   
93233      ORAL         2010-09-27              NaT                   NDA   
93234      ORAL         2010-09-27              NaT                   NDA   
93235      ORAL         2010-09-27              NaT                   NDA   
93236   TOPICAL         2006-03-13              NaT                   NDA   
93237   TOPICAL         2013-11-14              NaT                   NDA   

      APPLICATIONNUMBER                                LABELERNAME  \
20            NDA021411                                         10   
21            NDA021427                                         10   
22            NDA020592                                         10   
23            NDA020592                                         10   
24            NDA020592                                         10   
...                 ...                                        ...   
93233         NDA050808  Valeant Pharmaceutica

In [ ]:
dfProd.drop_duplicates(subset=['PRODUCTID'], keep="first", inplace=False)

**A partir d'ici c'est du brouillon**

In [36]:
for index, row in dfProd.iterrows():
    dfProd.loc[index,'PRODUCTID'] = row['PRODUCTNDC']
    

In [37]:
segments = dfProd['PRODUCTNDC'].astype(str).apply(lambda x: '-'.join(x.split('-')[:2])if pd.notna(x) else pd.NaN)
segments 

20       0002-3251
21       0002-3270
22       0002-4112
23       0002-4115
24       0002-4116
           ...    
93233    99207-465
93234    99207-466
93235    99207-467
93236    99207-525
93237    99207-850
Name: PRODUCTNDC, Length: 89372, dtype: object

In [38]:
df = pd.DataFrame({'col1': ['1234-234-43', '5678-456-78', '9012-678-90',np.NaN]})

# extract the substring before the second hyphen for each value in col1
df['new_col'] = df['col1'].apply(lambda x: '-'.join(x.split('-')[:2])if pd.notna(x) else np.NaN)
df

col1   new_col
0  1234-234-43  1234-234
1  5678-456-78  5678-456
2  9012-678-90  9012-678
3          NaN       NaN

In [39]:
#nombre de valeurs en commun
common_strings = set(dfProd['PRODUCTID'].str.lower()) & set(dfPack['PRODUCTID'].str.lower())
print(len(common_strings))
common_strings = set(dfProd['PRODUCTNDC'].str.lower()) & set(dfPack['PRODUCTNDC'].str.lower())
print(len(common_strings))
common_strings = set(dfProd['STARTMARKETINGDATE']) & set(dfPack['STARTMARKETINGDATE'])
print(len(common_strings))
common_strings = set(dfProd['ENDMARKETINGDATE']) & set(dfPack['ENDMARKETINGDATE'])
print(len(common_strings))
common_strings = set(dfProd['NDC_EXCLUDE_FLAG'].str.lower()) & set(dfPack['NDC_EXCLUDE_FLAG'].str.lower())
print(common_strings)

0
86445
6959
735
{'n'}


In [40]:
related_features = []
for i in dfProd.columns :
    if i in dfPack.columns:
        related_features.append(i)
print(f"Les deux tables ont ces attributs en commun :",related_features)

Les deux tables ont ces attributs en commun : ['PRODUCTID', 'PRODUCTNDC', 'STARTMARKETINGDATE', 'ENDMARKETINGDATE', 'NDC_EXCLUDE_FLAG']
